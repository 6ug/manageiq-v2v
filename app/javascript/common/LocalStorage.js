export const getLocalStorageState = key => {
  try {
    const serializedState = localStorage.getItem(key);
    if (serializedState === null) {
      return undefined;
    }
    return JSON.parse(serializedState);
  } catch (err) {
    return undefined;
  }
};

export const getLocalStorageNumericState = (key, defaultValue) => {
  const value = getLocalStorageState(key);
  if (!isNaN(value)) {
    return parseInt(value, 10);
  } else if (defaultValue) {
    saveLocalStorageState(key, defaultValue);
    return defaultValue;
  }
  return undefined;
};

export const saveLocalStorageState = (key, value) => {
  try {
    const serializedState = JSON.stringify(value);
    localStorage.setItem(key, serializedState);
  } catch (err) {
    console.err(err);
  }
};

export const LOCAL_STORAGE_KEYS = {
  V2V_TRANSFORMATION_MAPPINGS: 'V2V_TRANSFORMATION_MAPPINGS',
  V2V_PLANS: 'V2V_PLANS',
  V2V_UI_POLL_INTERVAL: 'V2V_UI_POLL_INTERVAL'
};
