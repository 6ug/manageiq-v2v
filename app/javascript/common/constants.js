export const MAX_LENGTH_NAMES = 24;
export const MAX_LENGTH_DESCRIPTIONS = 128;

const maxLengthIs = `Maximum length is %s characters.`;
const requiredMaxLengthIs = `Required. ${maxLengthIs}`;
const youHaveReachedMax = `You have reached the maximum length of %s characters.`;

export const validation = {
  name: {
    help: sprintf(__(maxLengthIs), MAX_LENGTH_NAMES),
    requiredMessage: sprintf(__(requiredMaxLengthIs), MAX_LENGTH_NAMES),
    maxLength: MAX_LENGTH_NAMES,
    maxLengthWarning: sprintf(__(youHaveReachedMax), MAX_LENGTH_NAMES)
  },
  description: {
    help: sprintf(__(maxLengthIs), MAX_LENGTH_DESCRIPTIONS),
    maxLength: MAX_LENGTH_DESCRIPTIONS,
    maxLengthWarning: sprintf(__(youHaveReachedMax), MAX_LENGTH_DESCRIPTIONS)
  }
};
