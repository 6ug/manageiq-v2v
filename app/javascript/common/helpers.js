// todo: expose this from patternfly-react instead?
export const bindMethods = (context, methods) => {
  methods.forEach(method => {
    context[method] = context[method].bind(context);
  });
};

export const noop = Function.prototype;

export const getDisplayName = Component =>
  Component.displayName || Component.name || 'Component';

// https://github.com/yahoo/react-intl/wiki/Upgrade-Guide#flatten-messages-object
export const flattenMessages = (nestedMessages, prefix = '') =>
  Object.keys(nestedMessages).reduce((messages, key) => {
    const value = nestedMessages[key];
    const prefixedKey = prefix ? `${prefix}.${key}` : key;

    if (typeof value === 'string') {
      messages[prefixedKey] = value;
    } else {
      Object.assign(messages, flattenMessages(value, prefixedKey));
    }

    return messages;
  }, {});
