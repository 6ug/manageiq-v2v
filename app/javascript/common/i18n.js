import { addLocaleData } from 'react-intl';

// hard wired to english for now
import en from 'react-intl/locale-data/en';

class I18n {
  constructor(locale, timezone) {
    this.fallbackIntl = !global.Intl;

    [this.locale] = locale.split('-');
    this.timezone = this.fallbackIntl ? 'UTC' : timezone;

    // this.loaded = false;
    // this.ready = this.init();

    // hard wired for now
    addLocaleData([...en]);
    this.loaded = true;
  }

  // todo: gather react-intl locale-data from the server asynchronously
  // async init() {
  //   await this.fetchIntl();
  //   addLocaleData(await import(`react-intl/locale-data/${this.locale}`));
  //   this.loaded = true;
  // }

  // async fetchIntl() {
  //   if (this.fallbackIntl) {
  //     global.Intl = await import('intl');
  //     await import(`intl/locale-data/jsonp/${this.locale}`);
  //   }
  // }
}

const [htmlElemnt] = document.getElementsByTagName('html');
const langAttr = htmlElemnt.getAttribute('lang') || 'en';
const timezoneAttr = htmlElemnt.getAttribute('data-timezone') || 'UTC';

const i18n = new I18n(langAttr, timezoneAttr);

export default i18n;
