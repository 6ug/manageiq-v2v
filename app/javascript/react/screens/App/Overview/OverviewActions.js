import URI from 'urijs';

import API, {
  globalMockMode,
  globalLocalStorageMode
} from '../../../../common/API';

import {
  getLocalStorageState,
  saveLocalStorageState,
  LOCAL_STORAGE_KEYS
} from '../../../../common/LocalStorage';

import {
  SHOW_MAPPING_WIZARD,
  SHOW_PLAN_WIZARD,
  HIDE_MAPPING_WIZARD,
  FETCH_V2V_TRANSFORMATION_MAPPINGS,
  FETCH_V2V_TRANSFORMATION_PLANS,
  CREATE_V2V_TRANSFORMATION_PLAN_REQUEST,
  V2V_FETCH_CLUSTERS,
  V2V_SET_MIGRATIONS_FILTER
} from './OverviewConstants';

import {
  requestTransformationMappingsData,
  createTransformationPlanRequestData
} from './overview.fixtures';
import { requestTransformationPlansData } from './overview.transformationPlans.fixtures';
import { requestClustersData } from './overview.clusters.fixtures';

const mockMode = globalMockMode;
const localStorageMode = globalLocalStorageMode;

export const showMappingWizardAction = () => dispatch => {
  dispatch({
    type: SHOW_MAPPING_WIZARD
  });
};

export const showPlanWizardAction = id => dispatch => {
  dispatch({
    type: SHOW_PLAN_WIZARD,
    payload: id
  });
};

const _getPlanFromLocalStorage = id => {
  const plans = getLocalStorageState(LOCAL_STORAGE_KEYS.V2V_PLANS);
  return plans.find(p => p.id === id);
};

const _createTransformationPlanRequestActionCreator = (
  url,
  planId
) => dispatch => {
  if (localStorageMode) {
    // in localStorageMode, we just want to modify the plan's pending tasks and
    // set them pending
    const plans = getLocalStorageState(LOCAL_STORAGE_KEYS.V2V_PLANS);
    const planLocalStorageIndex = plans.findIndex(p => p.id === planId);
    const plan = plans[planLocalStorageIndex];

    // todo: handle retry as well as not started
    if (plan.pending_tasks) {
      // create a request with the previously created pending_tasks
      const request = {};
      request.description = plan.name;
      request.approval_state = 'approved';
      request.status = 'pending';
      request.created_on = new Date().toString();
      request.miq_request_tasks = plan.pending_tasks;
      plan.miq_requests.push(request);
      plan.pending_tasks = [];
      plans[planLocalStorageIndex] = plan;
    }
    saveLocalStorageState(LOCAL_STORAGE_KEYS.V2V_PLANS, plans);

    dispatch({
      type: `${CREATE_V2V_TRANSFORMATION_PLAN_REQUEST}_FULFILLED`,
      payload: { data: plan }
    });
  }
  return dispatch({
    type: CREATE_V2V_TRANSFORMATION_PLAN_REQUEST,
    payload: {
      promise: API.post(url, { action: 'order' }),
      data: url
    }
  }).catch(error => {
    if (mockMode) {
      dispatch({
        type: `${CREATE_V2V_TRANSFORMATION_PLAN_REQUEST}_FULFILLED`,
        payload: createTransformationPlanRequestData.response
      });
    }
  });
};

export const createTransformationPlanRequestAction = (url, planId) => {
  const uri = new URI(url);
  return _createTransformationPlanRequestActionCreator(uri.toString(), planId);
};

const _getTransformationMappingsActionCreator = url => dispatch => {
  if (localStorageMode) {
    const mappings = getLocalStorageState(
      LOCAL_STORAGE_KEYS.V2V_TRANSFORMATION_MAPPINGS
    );
    dispatch({
      type: `${FETCH_V2V_TRANSFORMATION_MAPPINGS}_FULFILLED`,
      payload: { data: { resources: mappings || [] } }
    });
  } else {
    dispatch({
      type: FETCH_V2V_TRANSFORMATION_MAPPINGS,
      payload: API.get(url)
    }).catch(error => {
      if (mockMode) {
        dispatch({
          type: `${FETCH_V2V_TRANSFORMATION_MAPPINGS}_FULFILLED`,
          payload: requestTransformationMappingsData.response
        });
      }
    });
  }
};

export const fetchTransformationMappingsAction = url => {
  const uri = new URI(url);
  return _getTransformationMappingsActionCreator(uri.toString());
};

const _getTransformationPlansActionCreator = url => dispatch => {
  if (localStorageMode) {
    const plans = getLocalStorageState(LOCAL_STORAGE_KEYS.V2V_PLANS);
    return dispatch({
      type: FETCH_V2V_TRANSFORMATION_PLANS,
      payload: Promise.resolve({ data: { resources: plans || [] } })
    });
  } else {
    dispatch({
      type: 'FETCH_V2V_TRANSFORMATION_PLANS',
      payload: API.get(url)
    }).catch(error => {
      if (mockMode) {
        dispatch({
          type: `${FETCH_V2V_TRANSFORMATION_PLANS}_FULFILLED`,
          payload: requestTransformationPlansData.response
        });
      }
    });
  }
};

export const fetchTransformationPlansAction = url => {
  const uri = new URI(url);
  return _getTransformationPlansActionCreator(uri.toString());
};

export const continueToPlanAction = id => dispatch => {
  dispatch({
    type: HIDE_MAPPING_WIZARD
  });
  dispatch({
    type: SHOW_PLAN_WIZARD,
    payload: { id }
  });
};

const _getClustersActionCreator = url => dispatch =>
  dispatch({
    type: `${V2V_FETCH_CLUSTERS}`,
    payload: API.get(url)
  }).catch(error => {
    if (mockMode) {
      dispatch({
        type: `${V2V_FETCH_CLUSTERS}_FULFILLED`,
        payload: requestClustersData.response
      });
    }
  });

export const fetchClustersAction = url => {
  const uri = new URI(url);
  return _getClustersActionCreator(uri.toString());
};

export const setMigrationsFilterAction = filter => ({
  type: V2V_SET_MIGRATIONS_FILTER,
  payload: filter
});
