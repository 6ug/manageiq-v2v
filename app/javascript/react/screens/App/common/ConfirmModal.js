import React from 'react';
import PropTypes from 'prop-types';
import { Modal, Button, Grid, noop } from 'patternfly-react';

const ConfirmModal = props => {
  const {
    icon,
    title,
    body,
    cancelButtonLabel,
    confirmButtonLabel,
    onCancel,
    onConfirm,
    ...otherProps
  } = props;
  return (
    <Modal {...otherProps} onHide={onCancel}>
      <Modal.Header>
        <Modal.CloseButton onClick={onCancel} />
        <Modal.Title>{title}</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        {!icon ? (
          body
        ) : (
          <Grid.Row className="show-grid">
            <Grid.Col md={2} className="text-right">
              {icon}
            </Grid.Col>
            <Grid.Col md={10}>{body}</Grid.Col>
          </Grid.Row>
        )}
      </Modal.Body>
      <Modal.Footer>
        <Button bsStyle="default" className="btn-cancel" onClick={onCancel}>
          {cancelButtonLabel}
        </Button>
        <Button bsStyle="primary" onClick={onConfirm}>
          {confirmButtonLabel}
        </Button>
      </Modal.Footer>
    </Modal>
  );
};

ConfirmModal.propTypes = {
  show: PropTypes.bool,
  title: PropTypes.string,
  icon: PropTypes.node,
  body: PropTypes.node,
  cancelButtonLabel: PropTypes.string,
  confirmButtonLabel: PropTypes.string,
  onCancel: PropTypes.func,
  onConfirm: PropTypes.func
};

ConfirmModal.defaultProps = {
  show: false,
  title: __('Confirm'),
  icon: null,
  body: <p>{__('Are you sure?')}</p>,
  cancelButtonLabel: __('Cancel'),
  confirmButtonLabel: __('Yes'),
  onCancel: noop,
  onConfirm: noop
};

export default ConfirmModal;
