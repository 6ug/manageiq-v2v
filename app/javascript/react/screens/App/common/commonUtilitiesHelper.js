const commonUtilitiesHelper = {
  getMostRecentEntityByCreationDate: entities =>
    entities.reduce((prev, current) => (prev.created_on > current.created_on ? prev : current)),

  groupBy: (items, key) =>
    items.reduce(
      (result, item) => ({
        ...result,
        [item[key]]: [...(result[item[key]] || []), item]
      }),
      {}
    ),

  flattenArray: (arrayItems, flattenedArray) =>
    arrayItems.map(task => task.map(arrayElement => flattenedArray.push(arrayElement)))
};

export default commonUtilitiesHelper;
