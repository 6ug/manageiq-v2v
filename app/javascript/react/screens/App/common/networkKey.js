/**
 * returns a unique network based on current UI model
 */
export const networkKey = network => `${network.id}\\${network.uid_ems}`;
