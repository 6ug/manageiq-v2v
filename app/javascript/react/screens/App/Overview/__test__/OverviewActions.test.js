import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { mockRequest, mockReset } from '../../../../../common/mockRequests';

import { fetchTransformationPlansAction } from '../OverviewActions';
import { requestTransformationPlansData } from '../overview.transformationPlans.fixtures';

const middlewares = [thunk, promiseMiddleware()];
const mockStore = configureMockStore(middlewares);
const store = mockStore({});

afterEach(() => {
  store.clearActions();
  mockReset();
});

describe('fetchTransformationPlansAction', () => {
  const { fetchTransformationPlansUrl } = requestTransformationPlansData;
  const request = { fetchTransformationPlansUrl };

  test('dispatches PENDING and FULFILLED actions', () => {
    mockRequest({
      ...request,
      status: 200
    });

    return store
      .dispatch(fetchTransformationPlansAction(fetchTransformationPlansUrl))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });

  test('dispatches PENDING and REJECTED actions', () => {
    mockRequest({
      ...request,
      status: 404
    });

    return store
      .dispatch(fetchTransformationPlansAction(fetchTransformationPlansUrl))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });
});
