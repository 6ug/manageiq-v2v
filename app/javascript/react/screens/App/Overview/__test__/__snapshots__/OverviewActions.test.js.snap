// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchTransformationPlansAction dispatches PENDING and FULFILLED actions 1`] = `
Array [
  Object {
    "type": "FETCH_V2V_TRANSFORMATION_PLANS_PENDING",
  },
  Object {
    "payload": Object {
      "config": Object {
        "data": undefined,
        "headers": Object {
          "Accept": "application/json, text/plain, */*",
          "X-Auth-Token": "",
          "X-Requested-With": "XMLHttpRequest",
        },
        "maxContentLength": -1,
        "method": "get",
        "params": Object {},
        "timeout": 0,
        "transformRequest": Object {
          "0": [Function],
        },
        "transformResponse": Object {
          "0": [Function],
        },
        "url": "/api/dummyTransformationPlans",
        "validateStatus": [Function],
        "xsrfCookieName": "XSRF-TOKEN",
        "xsrfHeaderName": "X-XSRF-TOKEN",
      },
      "data": null,
      "headers": undefined,
      "status": 200,
    },
    "type": "FETCH_V2V_TRANSFORMATION_PLANS_FULFILLED",
  },
]
`;
